apm-sniffer/optional-reporter-plugins/kafka-reporter-plugin/src/main/java/org/apache/skywalking/apm/agent/core/kafka/KafkaServiceManagementServiceClient.java

```diff
@@ -85,28 +87,32 @@ public void boot() {
         ), 0, Config.Collector.HEARTBEAT_PERIOD, TimeUnit.SECONDS);
-
-        InstanceProperties instance = InstanceProperties.newBuilder()
-                                                        .setService(Config.Agent.SERVICE_NAME)
-                                                        .setServiceInstance(Config.Agent.INSTANCE_NAME)
-                                                        .addAllProperties(OSUtil.buildOSInfo(
-                                                            Config.OsInfo.IPV4_LIST_SIZE))
-                                                        .addAllProperties(SERVICE_INSTANCE_PROPERTIES)
-                                                        .build();
-        producer.send(new ProducerRecord<>(topic, TOPIC_KEY_REGISTER + instance.getServiceInstance(), Bytes.wrap(instance.toByteArray())));
-        producer.flush();
     }
 
     @Override
     public void run() {
-        InstancePingPkg ping = InstancePingPkg.newBuilder()
-                                              .setService(Config.Agent.SERVICE_NAME)
-                                              .setServiceInstance(Config.Agent.INSTANCE_NAME)
-                                              .build();
-        if (LOGGER.isDebugEnable()) {
-            LOGGER.debug("Heartbeat reporting, instance: {}", ping.getServiceInstance());
+        if (Math.abs(sendPropertiesCounter.getAndAdd(1)) % Config.Collector.TIMES_SEND_PROPERTIES == 0) {
+            InstanceProperties instance = InstanceProperties.newBuilder()
+                                                            .setService(Config.Agent.SERVICE_NAME)
+                                                            .setServiceInstance(Config.Agent.INSTANCE_NAME)
+                                                            .addAllProperties(OSUtil.buildOSInfo(
+                                                                Config.OsInfo.IPV4_LIST_SIZE))
+                                                            .addAllProperties(SERVICE_INSTANCE_PROPERTIES)
+                                                            .build();
+            producer.send(new ProducerRecord<>(topic, TOPIC_KEY_REGISTER + instance.getServiceInstance(),
+                                               Bytes.wrap(instance.toByteArray())
+            ));
+            producer.flush();
+        } else {
+            InstancePingPkg ping = InstancePingPkg.newBuilder()
+                                                  .setService(Config.Agent.SERVICE_NAME)
+                                                  .setServiceInstance(Config.Agent.INSTANCE_NAME)
+                                                  .build();
+            if (LOGGER.isDebugEnable()) {
+                LOGGER.debug("Heartbeat reporting, instance: {}", ping.getServiceInstance());
+            }
+            producer.send(new ProducerRecord<>(topic, ping.getServiceInstance(), Bytes.wrap(ping.toByteArray())));
```

Comment:
Why do you move this `ping` into the `else` block? This should be as what it was.

Reply:
The KafkaServiceManagementServiceClient need sends the instance properties to the backend every `collector.heartbeat_period * collector.times_send_properties` seconds

Reply:
@kezhenxu94 KafkaServiceManagementServiceClient also need to send the instance properties.

Reply:
I think the question is, ping should be sent every time, no `else`. You should not change this part of the original codes.